### PR TITLE
FollowUpTransformer and ObservationPeriodTransformer transformed in d…

### PR DESCRIPTION
**Pull request by Anfor**
This modification has an optimization as well, I don't have used dispensations dataset in  FollowUpTransformer class because later analyze it, I don't have found any use of it.

### CODE REVIEW

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import java.sql.Timestamp
+import org.joda.time.{DateTime, DateTimeZone}
+import org.apache.spark.sql.Dataset
+import fr.polytechnique.cmap.cnam.etl.events.{Event, FollowUp}
+import fr.polytechnique.cmap.cnam.etl.transformers.follow_up.Columns.EndReasons
+
+object FollowUpTransformerUtilities {
+
+
+  case class PatientsValues(
```
Comment (dsun0720) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: The name "PatientsValues" is not clear. I think we can call it PatientsFollowup or PatientsDates. What do you think ?

Comment (dsun0720) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: PatientDates is better ? what do you think ?

Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: It's seems ok for me

**Comment by dsun0720** [comment]
And I suggest you apply the functor style that youcef did in this transformer

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
+package fr.polytechnique.cmap.cnam.etl.transformers.follow_up
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: As this class has many problems, I would suggest the following implementation:
```scala
package fr.polytechnique.cmap.cnam.etl.transformers.follow_up

import java.sql.Timestamp
import org.joda.time.{DateTime, DateTimeZone}
import org.apache.spark.sql.Dataset
import fr.polytechnique.cmap.cnam.etl.events.{Event, FollowUp}
import fr.polytechnique.cmap.cnam.etl.transformers.follow_up.Columns.EndReasons

object FollowUpTransformerUtilities {


  case class PatientDates(
    patientID: String,
    deathDate: Option[Timestamp],
    followUpStart: Option[Timestamp],
    observationEnd: Option[Timestamp])


  case class TrackLossDate(
    patientID: String,
    trackloss: Option[Timestamp])

  sealed trait FollowUpEndReason extends Timestamp{
    val reason: String
  }

  class Death extends FollowUpEndReason {
    val reason = EndReasons.Death.toString
  }

  class Disease extends FollowUpEndReason {
    override val reason: String = EndReasons.Disease.toString
  }

  class TrackLoss extends FollowUpEndReason {
    override val reason: String = EndReasons.Trackloss.toString
  }

  class ObservationEnd extends FollowUpEndReason {
    override val reason: String = EndReasons.ObservationEnd.toString
  }


  val correctedStart: (Timestamp, Option[Timestamp], Int) => Option[Timestamp] =
    (start: Timestamp, end: Option[Timestamp], delayMonths: Int) => {
      val st = Timestamp.valueOf(
        new DateTime(start, DateTimeZone.UTC).plusMonths(delayMonths)
          .toString("YYYY-MM-dd HH:mm:ss")
      )
      val corrected: Option[Timestamp] = if (st.before(end.get)) Some(st) else None
      corrected
    }

  val tracklossDateCorrected: (Timestamp, Timestamp) => Option[Timestamp] =
    (start: Timestamp, followUpStart: Timestamp) => {
      val corrected: Option[Timestamp] = if (start.after(followUpStart)) Some(start) else None
      corrected
    }
  

  val endReason: (Option[Timestamp], Option[Timestamp], Option[Timestamp], Option[Timestamp]) => Option[FollowUpEndReason] =
    (
      observationEnd: Option[Timestamp],
      deathDate: Option[Timestamp],
      trackloss: Option[Timestamp], 
      firstTargetDiseaseDate: Option[Timestamp]) => {

      Seq(
        trackloss.map(_.asInstanceOf[TrackLoss]),
        deathDate.map(_.asInstanceOf[Death]),
        observationEnd.map(_.asInstanceOf[ObservationEnd]),
        firstTargetDiseaseDate.map(_.asInstanceOf[Disease])
      ).min
    }


  implicit class FollowUpDataset(followups: Dataset[Event[FollowUp]]) {
    def cleanFollowUps(): Dataset[Event[FollowUp]] = followups.filter(e => e.start.before(e.end.get))
  }

}
```

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/observation/ObservationPeriodTransformer.scala [hunk 3]
```diff
@@ -3,62 +3,37 @@
...
-  def transform(events: Dataset[Event[AnyEvent]]): Dataset[Event[ObservationPeriod]] = {
+    val studyEnd = config.studyEnd
 
     import events.sqlContext.implicits._
 
-    val observation = events.toDF
-      .withObservationStart
-      .withObservationEnd
-      .dropDuplicates(Seq("patientID"))
+    val observation: Dataset[Event[ObservationPeriod]] = events.filter(
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/observation/ObservationPeriodTransformer.scala]: No need to store t in a val. Just use the statement as a terminal clause.

File: src/main/scala/fr/polytechnique/cmap/cnam/study/pioglitazone/PioglitazoneMain.scala [hunk 4]
```diff
@@ -18,7 +18,7 @@ import fr.polytechnique.cmap.cnam.etl.patients.Patient
 import fr.polytechnique.cmap.cnam.etl.sources.Sources
 import fr.polytechnique.cmap.cnam.etl.transformers.exposures.ExposuresTransformer
 import fr.polytechnique.cmap.cnam.etl.transformers.follow_up.FollowUpTransformer
-import fr.polytechnique.cmap.cnam.etl.transformers.follow_up.FollowUpTransformer.FollowUpDataset
+import fr.polytechnique.cmap.cnam.etl.transformers.follow_up.FollowUpTransformerUtilities.FollowUpDataset
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/study/pioglitazone/PioglitazoneMain.scala]: This is bad. Utilities shoudn't be involved in high level classes like this one.

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala [hunk 5]
```diff
@@ -2,14 +2,25 @@
...
 import org.apache.spark.sql.Dataset
-import org.apache.spark.sql.functions.lit
-import org.apache.spark.sql.types.TimestampType
 import fr.polytechnique.cmap.cnam.SharedContext
 import fr.polytechnique.cmap.cnam.etl.events._
 import fr.polytechnique.cmap.cnam.etl.patients.Patient
+import fr.polytechnique.cmap.cnam.etl.transformers.follow_up.FollowUpTransformerUtilities.{PatientDates, endReason, tracklossDate, tracklossDateCorrected}
 import fr.polytechnique.cmap.cnam.util.functions.makeTS
 
+case class inputDS(
```
Comment (ysebiat) [inline src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala]: The naming is bad & why is this?

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala [hunk 6]
```diff
@@ -2,14 +2,25 @@
...
+  patientID: String,
+  observationStart: Timestamp,
+  observationEnd: Option[Timestamp])
+
+case class resultDS(
```
Comment (ysebiat) [inline src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala]: Bad naming & why this?

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala [hunk 7]
```diff
@@ -2,119 +2,100 @@
...
+        PatientDates(
+          e._1.patientID,
+          e._1.deathDate,
+          correctedStart(e._2.start, e._2.end, delayMonths),
+          e._2.end
+        )
+      }
+      .filter(e => e.followUpStart.nonEmpty)
+      .groupBy(col(PatientID))
+      .agg(
```
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala]: I takes the min date of death date, the min follow up date and the min of observation date, to get all the min dates of the patient.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala [hunk 8]
```diff
@@ -2,119 +2,100 @@
...
+    val disease = config.outcomeName.getOrElse(None).toString
 
-      data.withColumn(TracklossDate, firstCorrectTrackloss)
-    }
+    val outcomesDisease: Dataset[Event[Outcome]] = outcomes
+      .filter(e => e.value.matches(s".*$disease.*"))
+      .groupBy(col(PatientID))
+      .agg(
+        min(Start).as(Start)
+      ).map(e => Outcome(e.getAs[String](PatientID), disease, e.getAs[Timestamp](Start)))
```
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala]: In this mapping outcome apply method doesn't resolve correctly.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+import java.sql.Timestamp
+import org.joda.time.{DateTime, DateTimeZone}
```
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: Joda is the standard to manipulate dates,to add the months to followUp Start is needed.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+
+  case class PatientDates(
+    patientID: String,
+    deathDate: Option[Timestamp],
+    followUpStart: Option[Timestamp],
+    observationEnd: Option[Timestamp])
+
+
+  case class tracklossDate(
```
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: ok, I change

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala (hunk 2 above)
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: I agree with you.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala [hunk 11]
```diff
@@ -2,119 +2,100 @@
...
-
-      val followUpStart = add_months(col(ObservationStart), followUpMonthsDelay).cast(TimestampType)
-      val correctedFollowUpStart = when(followUpStart < col(ObservationEnd), followUpStart)
-
-      data.withColumn(FollowUpStart, min(correctedFollowUpStart).over(window))
-    }
-
-    def withTrackloss: DataFrame = {
-      val window = Window.partitionBy(PatientID)
+    val tracklossDates: Dataset[tracklossDate] = patientDates
```
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala]: I change it.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala [hunk 12]
```diff
@@ -2,119 +2,100 @@
...
+      .joinWith(tracklosses, tracklosses.col(PatientID) === patientDates.col(PatientID))
+      .map(e => tracklossDate(e._2.patientID, tracklossDateCorrected(e._2.start, e._1.followUpStart.get)))
+      .filter(e => e.trackloss.nonEmpty)
+      .groupBy(col(PatientID))
+      .agg(
+        min(TracklossDate).as(TracklossDate)
+      )
+      .map(e => tracklossDate(e.getAs[String](PatientID), Option(e.getAs[Timestamp](TracklossDate))))
```
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala]: I don't mind to change it, but this implementation throw me lot of errors

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala [hunk 13]
```diff
@@ -2,119 +2,100 @@
...
 
-      val firstCorrectTrackloss = min(
-        when(col(Category) === Trackloss.category && (col(Start) > col(FollowUpStart)), col(Start))
-      ).over(window)
+    val disease = config.outcomeName.getOrElse(None).toString
 
-      data.withColumn(TracklossDate, firstCorrectTrackloss)
-    }
+    val outcomesDisease: Dataset[Event[Outcome]] = outcomes
+      .filter(e => e.value.matches(s".*$disease.*"))
```
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformer.scala]: To get only the values of the studied disease

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala (hunk 2 above)
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: I don't mind to change it, but this implementation throw me lot of errors

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+
+  val minFollowupEnd: (Option[Timestamp], Option[Timestamp], Option[Timestamp]) => Option[Timestamp] =
+    (
+      tracklossDate: Option[Timestamp],
+      deathDate: Option[Timestamp],
+      observationEnd: Option[Timestamp]) => {
+
+      val times: Seq[Option[Timestamp]] = Seq(tracklossDate, deathDate, observationEnd).filter(e => e.isDefined)
```
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: Yes this is the goal.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/observation/ObservationPeriodTransformer.scala (hunk 3 above)
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/observation/ObservationPeriodTransformer.scala]: ok

File: src/main/scala/fr/polytechnique/cmap/cnam/study/pioglitazone/PioglitazoneMain.scala (hunk 4 above)
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/study/pioglitazone/PioglitazoneMain.scala]: ok

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala (hunk 5 above)
Comment (Anfor) [reply src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala]: Dataset in the class is naming input, this is the reason for the name

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala (hunk 6 above)
Comment (Anfor) [reply src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala]: Is the result dataset.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (ysebiat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: We use another library. It is already in the build. Check it out.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala (hunk 2 above)
Comment (ysebiat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: What are the thrown errors?

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala (hunk 5 above)
Comment (ysebiat) [reply src/test/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerSuite.scala]: I dont understand what you said. You can explain more please?

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: Do you mean java.sql?

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala (hunk 2 above)
Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: For instance:

Error:(25, 23) overloaded method constructor Timestamp with alternatives:
  (x$1: Long)java.sql.Timestamp <and>
  (x$1: Int,x$2: Int,x$3: Int,x$4: Int,x$5: Int,x$6: Int,x$7: Int)java.sql.Timestamp
 cannot be applied to ()
  class Death extends FollowUpEndReason {

Comment (Anfor) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: I'm trying to implement it and I found this error as well Error:(72, 9) No implicit Ordering defined for Option[fr.polytechnique.cmap.cnam.etl.transformers.follow_up.FollowUpTransformerUtilities.FollowUpEndReason].
      ).min 

I can't use implicit ordering with Option[FollowUpEndReason]

Comment (ysebiat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: For the first, just add the date as an argument and define the case class.

As for the second, define an implicit value in the scope that defines the Ordering for the type. I taught that inheriting from Timestamp would have defined the Ordering, which is not the case.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 15]
```diff
@@ -16,77 +15,80 @@ object FollowUpTransformerUtilities {
...
+    }
 
-  implicit class FollowUpDataset(followups: Dataset[Event[FollowUp]]) {
-    def cleanFollowUps(): Dataset[Event[FollowUp]] = followups.filter(e => e.start.before(e.end.get))
+  def endReason(
+    death: DeathReason,
+    disease: DiseaseReason,
+    trackloss: TrackLossReason,
+    observation: ObservationEndReason): FollowUpEnd = {
+    println(Seq(death, disease, trackloss, observation))
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: You don't need this here.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 16]
```diff
@@ -16,77 +15,80 @@ object FollowUpTransformerUtilities {
...
 
-      val endReason: String = followUpEnd match {
-        case death if (death.equals(deathDate)) => EndReasons.Death.toString
-        case disease if (disease.equals(firstTargetDiseaseDate)) => EndReasons.Disease.toString
-        case track if (track.equals(trackloss)) => EndReasons.Trackloss.toString
-        case obsEnd if (obsEnd.equals(observationEnd)) => EndReasons.ObservationEnd.toString
-      }
+  }
+
+  case class DeathReason(
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: Good.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 17]
```diff
@@ -16,77 +15,80 @@ object FollowUpTransformerUtilities {
...
     }
 
+  val tracklossDateCorrected: (Timestamp, Timestamp) => Option[Timestamp] =
+    (start: Timestamp, followUpStart: Timestamp) => {
+      if (start.after(followUpStart)) Some(start) else None
+    }
 
-  implicit class FollowUpDataset(followups: Dataset[Event[FollowUp]]) {
-    def cleanFollowUps(): Dataset[Event[FollowUp]] = followups.filter(e => e.start.before(e.end.get))
+  def endReason(
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: Great.

File: src/main/scala/fr/polytechnique/cmap/cnam/study/pioglitazone/PioglitazoneMain.scala [hunk 18]
```diff
@@ -302,7 +301,8 @@ object PioglitazoneMain extends Main {
       } else {
         firstFilterResult
       }
-      val cleanFollowUps = followups.cleanFollowUps() // Keep only followups for which start < stop
+      val cleanFollowUps = followups
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/study/pioglitazone/PioglitazoneMain.scala]: Perfect.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala [hunk 19]
```diff
@@ -16,77 +15,80 @@ object FollowUpTransformerUtilities {
...
+  abstract sealed class FollowUpEndReason {
+    val reason: String
+    val date: Option[Timestamp]
 
-  val tracklossDateCorrected: (Timestamp, Timestamp) => Option[Timestamp] =
-    (start: Timestamp, followUpStart: Timestamp) => {
-      val corrected: Option[Timestamp] = if (start.after(followUpStart)) Some(start) else None
-      corrected
-    }
+    def compare(that: FollowUpEndReason): Int = {
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/follow_up/FollowUpTransformerUtilities.scala]: Very good.

**Pull request merged**